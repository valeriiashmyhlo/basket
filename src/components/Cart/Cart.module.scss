.cartItem {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
};

.bold {
  font-weight: 600;
};
