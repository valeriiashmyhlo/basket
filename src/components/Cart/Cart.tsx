import * as React from "react";
import { Button, Table } from "react-bootstrap";
import { CartItem, Item } from "../../types";
import styles from "./Cart.module.scss";
import { roundToPrecision } from "../../utils/utils";

const Cart: React.FC<{
  items: CartItem[];
  onDeleteItem: (id: string) => void;
}> = ({ items, onDeleteItem }) => {
  if (items.length === 0) {
    return <div>Your cart is empty</div>;
  }

  const { total, savings: savingsTotal } = items.reduce(
    (a, c) => {
      const { total, savings } = c.item.priceRule(c.quantity);
      return {
        total: roundToPrecision(a.total + total, 2),
        savings: roundToPrecision(a.savings + savings, 2),
      };
    },
    { total: 0, savings: 0 }
  );

  const getItemsList = (items: CartItem[]): Item[] => {
    const list: Item[] = [];
    for (const item of items) {
      for (let i = 0; i < item.quantity; i++) {
        list.push(item.item);
      }
    }
    return list;
  };

  const subTotal = roundToPrecision(total + savingsTotal, 2);

  return (
    <Table striped size="sm">
      <thead>
        <tr>
          <th colSpan={4}>Title</th>
          <th>Price</th>
        </tr>
      </thead>
      <tbody>
        {getItemsList(items).map((item, i) => (
          <tr key={i}>
            <td>{item.name}</td>
            <td>{item.unitsPerItem}</td>
            <td>
              {item.pricePerUnit}/{item.unit}
            </td>
            <td>
              <Button size="sm" onClick={() => onDeleteItem(item.id)}>
                Remove
              </Button>
            </td>
            <td>{roundToPrecision(item.priceRule(1).total, 2)}</td>
          </tr>
        ))}
        <tr>
          <td colSpan={4} className={styles.bold}>Sub-total:</td>
          <td>£{subTotal}</td>
        </tr>
        <tr>
          <td colSpan={5} className={styles.bold}>Savings</td>
        </tr>
        {items.map(({ item, quantity }) =>
          item.priceRule(quantity).savings ? (
            <tr key={item.id}>
              <td colSpan={4}>{item.priceRuleText}</td>
              <td>-{roundToPrecision(item.priceRule(quantity).savings, 2)}</td>
            </tr>
          ) : null
        )}
        <tr>
          <td colSpan={4} className={styles.bold}>Total Savings:</td>
          <td>{-savingsTotal}</td>
        </tr>
        <tr>
          <td colSpan={4} className={styles.bold}>Total to Pay:</td>
          <td>£{total}</td>
        </tr>
      </tbody>
    </Table>
  );
};

export { Cart };
