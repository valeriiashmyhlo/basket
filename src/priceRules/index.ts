export { DefaultRule } from './DefaultRule/DefaultRule';
export { NForPrice } from './NForPrice/NForPrice';
export { NForM } from './NForM/NForM';