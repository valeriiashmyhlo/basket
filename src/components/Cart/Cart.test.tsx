import React from "react";
import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Cart } from "./Cart";
import { buildNForM, buildNForPrice } from "../../priceRules";

describe(`<Cart /> tests`, () => {
  const item1 = {
    item: {
      id: "1",
      name: "Face Mask",
      pricePerUnit: 2.5,
      unit: "each",
      unitsPerItem: 1,
      priceRule: buildNForPrice(2, 4),
      priceRuleText: "Face Masks 2 for Â£4",
    },
    quantity: 2,
  };
  const item2 = {
    item: {
      id: "2",
      name: "Toilet Paper",
      pricePerUnit: 0.65,
      unit: "roll",
      unitsPerItem: 1,
      priceRule: buildNForM(6, 5),
      priceRuleText: "Toilet Paper 6 for 5",
    },
    quantity: 2,
  };
  it("should render CartList with correct props", () => {
    const { container } = render(
      <Cart items={[item1, item2]} onDeleteItem={() => jest.fn()} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should apply correct rule", () => {
    const { container } = render(
      <Cart items={[item1]} onDeleteItem={() => jest.fn()} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call onDeleteItem", () => {
    const onDelete = jest.fn();
    const { getByText } = render(
      <Cart items={[item2]} onDeleteItem={onDelete} />
    );
    fireEvent.click(getByText("Remove"));
    expect(onDelete).toHaveBeenCalledTimes(1);
  });
});
