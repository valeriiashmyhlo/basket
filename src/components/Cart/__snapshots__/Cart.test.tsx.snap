// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> tests should apply correct rule 1`] = `
<table
  class="table table-sm table-striped"
>
  <thead>
    <tr>
      <th
        colspan="4"
      >
        Title
      </th>
      <th>
        Price
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Face Mask
      </td>
      <td>
        1
      </td>
      <td>
        2.5
        /
        each
      </td>
      <td>
        <button
          class="btn btn-primary btn-sm"
          type="button"
        >
          Remove
        </button>
      </td>
      <td>
        2.5
      </td>
    </tr>
    <tr>
      <td>
        Face Mask
      </td>
      <td>
        1
      </td>
      <td>
        2.5
        /
        each
      </td>
      <td>
        <button
          class="btn btn-primary btn-sm"
          type="button"
        >
          Remove
        </button>
      </td>
      <td>
        2.5
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="4"
      >
        Sub-total:
      </td>
      <td>
        £
        5
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="5"
      >
        Savings
      </td>
    </tr>
    <tr>
      <td
        colspan="4"
      >
        Face Masks 2 for £4
      </td>
      <td>
        -
        1
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="4"
      >
        Total Savings:
      </td>
      <td>
        -1
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="4"
      >
        Total to Pay:
      </td>
      <td>
        £
        4
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Cart /> tests should render CartList with correct props 1`] = `
<table
  class="table table-sm table-striped"
>
  <thead>
    <tr>
      <th
        colspan="4"
      >
        Title
      </th>
      <th>
        Price
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Face Mask
      </td>
      <td>
        1
      </td>
      <td>
        2.5
        /
        each
      </td>
      <td>
        <button
          class="btn btn-primary btn-sm"
          type="button"
        >
          Remove
        </button>
      </td>
      <td>
        2.5
      </td>
    </tr>
    <tr>
      <td>
        Face Mask
      </td>
      <td>
        1
      </td>
      <td>
        2.5
        /
        each
      </td>
      <td>
        <button
          class="btn btn-primary btn-sm"
          type="button"
        >
          Remove
        </button>
      </td>
      <td>
        2.5
      </td>
    </tr>
    <tr>
      <td>
        Toilet Paper
      </td>
      <td>
        1
      </td>
      <td>
        0.65
        /
        roll
      </td>
      <td>
        <button
          class="btn btn-primary btn-sm"
          type="button"
        >
          Remove
        </button>
      </td>
      <td>
        0.65
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="4"
      >
        Sub-total:
      </td>
      <td>
        £
        5.65
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="5"
      >
        Savings
      </td>
    </tr>
    <tr>
      <td
        colspan="4"
      >
        Face Masks 2 for £4
      </td>
      <td>
        -
        1
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="4"
      >
        Total Savings:
      </td>
      <td>
        -1
      </td>
    </tr>
    <tr>
      <td
        class="bold"
        colspan="4"
      >
        Total to Pay:
      </td>
      <td>
        £
        4.65
      </td>
    </tr>
  </tbody>
</table>
`;
